Gitlab support fixes

Benaiah opened the pull request:
Fixes several bugs with GitLab support in git-gateway:

- Adds `HEAD` requests to the allowed CORS methods.
- Adds `"Private-Token"` to the allowed CORS headers.
- Constructs a URL by setting `r.URL.Opaque` instead of by setting `r.URL.Path` to prevent literal `%2F`s in URLs, which are required by the GitLab API, from being turned into literal `/` characters by the Go stdlib. (see [this github issue](https://github.com/golang/go/issues/6658) and [this commit](https://github.com/golang/go/commit/874a605af0764a8f340c3de65406963f514e21bc), noting that `git-gateway` is running on `go1.8` if I understand correctly from the `.travis.yml`).
- Removes the `Authorization` header from GitLab requests, which causes 401s when used with private access tokens ([example](https://gitlab.com/gitlab-com/support-forum/issues/2722)), and replaces it with the `Private-Token` header. (The current WIP Netlify UI has a box for a token to be pasted, so I assume this is supposed to be a private access token, not an OAuth token - which would use the `Authorization` header).
- Rewrites absolute links in the GitLab `Link` header to be relative, so they're requested from git-gateway instead of the proxied API.
- Adds a default value for the GitLab API endpoint.

rybit commented on api/gitlab.go:
```diff
@@ -110,13 +119,40 @@ func (gl *GitLabGateway) authenticate(w http.ResponseWriter, r *http.Request) er
+	finalLinkEntries := []string{}
+	for _, linkEntry := range linkEntries {
+		linkAndRel := strings.Split(strings.TrimSpace(linkEntry), ";")
+		link := proxyAPIURL + strings.TrimPrefix(gitlabLinkRegex.FindStringSubmatch(linkAndRel[0])[1], endpointAPIURL)
```
you're indexing into matches, are we 100% sure that these regexes will not fail (odd inputs)?

Benaiah replied:
I added length-checking before match access to ensure that it's working as expected - the link entry is simply returned unchanged if something goes wrong (this is per-link, so the other links in the `Link` header can still be rewritten if one doesn't match what we expect).

rybit commented:
also if you want to update the Go version, it might be worth it right now.

calavera commented on api/gitlab.go:
```diff
@@ -110,13 +119,40 @@ func (gl *GitLabGateway) authenticate(w http.ResponseWriter, r *http.Request) er
+
+func rewriteGitlabLinks(linkHeader, endpointAPIURL, proxyAPIURL string) string {
+	linkEntries := strings.Split(linkHeader, ",")
+	finalLinkEntries := []string{}
```
you can save some allocations by creating the slice with fixed capacity:
```
finalLinkEntries := make([]string, len(linkEntries), len(linkEntries))
```

verythorough commented:
It looks like this requires a docs update. At this point, I think that means some minor changes to the README and an update to `example.env`

Benaiah commented:
@rybit @calavera addressed your review comments - could you re-review? @rybit upgrading Go sounds great, but I'm not sure I know the whole process for doing so for this repo.

rybit commented on api/gitlab.go:
```diff
@@ -110,13 +119,58 @@ func (gl *GitLabGateway) authenticate(w http.ResponseWriter, r *http.Request) er
+	linkEntries := strings.Split(linkHeader, ",")
+	finalLinkEntries := make([]string, len(linkEntries), len(linkEntries))
+	for _, linkEntry := range linkEntries {
+		finalLinkEntries = append(finalLinkEntries, rewriteGitlabLinkEntry(linkEntry, endpointAPIURL, proxyAPIURL))
```
I think that with the pre-allocation this will cause it to have a bunch of blanks. You'll want to use the index. 

https://play.golang.org/p/5a5P3ln5jZe

Benaiah replied:
Addressed, thanks for the catch.
